source/common/filesystem/win32/watcher_impl.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <concurrent_queue.h>
+
+#include <codecvt>
+#include <cstdint>
+#include <list>
+#include <locale>
+#include <string>
+#include <unordered_map>
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/filesystem/watcher.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/fmt.h"
+#include "common/common/logger.h"
+#include "common/common/thread_impl.h"
+#include "common/filesystem/filesystem_impl.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+class WatcherImpl : public Watcher, Logger::Loggable<Logger::Id::file> {
+public:
+  WatcherImpl(Event::Dispatcher& dispatcher);
+  ~WatcherImpl();
+
+  // Filesystem::Watcher
+  void addWatch(const std::string& path, uint32_t events, OnChangedCb cb) override;
+
+private:
+  static void issueFirstRead(ULONG_PTR param);
+  static void directoryChangeCompletion(DWORD err, DWORD num_bytes, LPOVERLAPPED overlapped);
+  static void endDirectoryWatch(SOCKET_FD sock, HANDLE hEvent);
+  void watchLoop();
+  void onDirectoryEvent();
+
+  struct FileWatch {
+    // store the wide character string for ReadDirectoryChangesW
+    std::wstring file_;
+    uint32_t events_;
+    OnChangedCb cb_;
+  };
+
+  typedef std::function<void(void)> CbClosure;
+
+  struct DirectoryWatch {
+    OVERLAPPED op_;
+    std::list<FileWatch> watches_;
+    HANDLE hDir_;
```

Comment:
nit: no hungarian notation please :)

Reply:
Note, the overlapped_.hEvent (formerly an illegible op_.hEvent) remains Hungarian, as a member of MS's struct. Otherwise addressed and several other watcher_impl variable names improved in the incoming commit.